pennylane_lightning/core/src/simulators/lightning_qubit/gates/cpu_kernels/GateImplementationsLM.hpp

```diff
@@ -483,6 +483,70 @@ class GateImplementationsLM : public PauliGenerator<GateImplementationsLM> {
         applyNCMultiQubitOp(arr, num_qubits, matrix, {}, {}, wires, inverse);
     }
 
+    /**
+     * @brief Apply a sparse matrix with controls to the statevector.
+     * @tparam PrecisionT Floating point precision of underlying statevector
+     * data
+     * @tparam index_type Index type
+     * @param arr Pointer to the statevector.
+     * @param num_qubits Number of qubits.
+     * @param row_map_ptr Pointer to the row map.
+     * @param col_idx_ptr Pointer to the columns indexes
+     * @param values_ptr Pointer to the values.
+     * @param controlled_wires Control wires.
+     * @param controlled_values Control values (false or true).
+     * @param wires Wires the gate applies to.
+     * @param inverse Indicate whether inverse should be taken.
+     */
+    template <class PrecisionT, class index_type = std::size_t>
+    static void applyNCMultiQubitSparseOp(
+        std::complex<PrecisionT> *arr, std::size_t num_qubits,
+        const index_type *row_map_ptr, const index_type *col_idx_ptr,
+        const std::complex<PrecisionT> *values_ptr,
+        const std::vector<std::size_t> &controlled_wires,
+        const std::vector<bool> &controlled_values,
+        const std::vector<std::size_t> &wires, bool inverse = false) {
+        constexpr std::size_t one{1U};
+        const std::size_t n_wires = wires.size();
+        const std::size_t dim = one << n_wires;
+
+        PL_ABORT_IF(inverse, "Inverse not supported for sparse ops.");
+
+        auto core_function = [dim, &row_map_ptr, &col_idx_ptr, &values_ptr](
+                                 std::complex<PrecisionT> *arr,
+                                 const std::vector<std::size_t> &indices,
+                                 const std::size_t offset) {
+            std::vector<std::complex<PrecisionT>> coeffs(dim);
+            for (std::size_t i = 0; i < dim; i++) {
+                coeffs[i] = arr[indices[i] + offset];
+            }
+            std::size_t count = 0;
+            for (index_type i = 0; i < dim; i++) {
```

Comment:
Could we use this pre-exiting function [in these kernels](https://github.com/PennyLaneAI/pennylane-lightning/blob/master/pennylane_lightning/core/src/simulators/lightning_qubit/utils/SparseLinAlg.hpp)?

Reply:
I don't see a clear path to do that without refactoring the `apply_Sparse_Matrix` function, and with that (possibly) needing to refactor the measure of sparse hamiltonians.
While the core_function logic in GateImplementationLM works by taking a pointer to `value_ptr` and acting upon that without returning anything, their equivalent in SparseLinAlg allocates and returns the `result` vector. Also, the support for SparseHamiltonians provided by the already existing Kernel does not support control wires as observed by the absence of any `coeffs` or the `indices` vector. If you think it is sensible, I can try refactoring the `apply_Sparse_Matrix`, it might be straightforward, but I don't have any time estimation for that.